test_runner/regress/test_tenant_size.py

```diff
@@ -0,0 +1,342 @@
+                cur.execute(
+                    f"UPDATE t0 SET i = -i WHERE i IN (SELECT i FROM t0 WHERE i > 0 LIMIT {batch_size})"
+                )
+                updated = cur.rowcount
+
+            if updated == 0:
+                break
+
+            current_lsn = wait_for_last_flush_lsn(env, pg, tenant_id, timeline_id)
+
+            size, _ = get_tenant_size(http_client, tenant_id)
+            assert size > collected_responses[-1][1]
+            collected_responses.append((current_lsn, size))
+
+        while True:
+            with pg.cursor() as cur:
+                cur.execute(f"DELETE FROM t0 WHERE i IN (SELECT i FROM t0 LIMIT {batch_size})")
+                deleted = cur.rowcount
+
+            if deleted == 0:
+                break
+
+            current_lsn = wait_for_last_flush_lsn(env, pg, tenant_id, timeline_id)
+
+            size, _ = get_tenant_size(http_client, tenant_id)
+            # even though rows have been deleted, the size should not decrease
+            assert size > collected_responses[-1][1]
+            collected_responses.append((current_lsn, size))
+
+        with pg.cursor() as cur:
+            cur.execute("DROP TABLE t0")
+
+        current_lsn = wait_for_last_flush_lsn(env, pg, tenant_id, timeline_id)
+
+        size, _ = get_tenant_size(http_client, tenant_id)
+
+        assert size > collected_responses[-1][1]
+        collected_responses.append((current_lsn, size))
+
+    for lsn, size in collected_responses:
+        log.info(f"collected: {lsn}, {size}")
+
+    env.pageserver.stop()
+    env.pageserver.start()
+
+    size_after = get_tenant_size(http_client, tenant_id)[0]
+
+    assert (
+        size_after == collected_responses[-1][1]
+    ), "size after restarting pageserver should not have changed"
+
+
+def test_get_tenant_size_with_multiple_branches(neon_env_builder: NeonEnvBuilder):
+    """
+    Reported size goes up while branches or rows are being added, goes down after removing branches.
+    """
+
+    neon_env_builder.pageserver_config_override = f"tenant_config={{compaction_period='1h', gc_period='1h', pitr_interval='0sec', gc_horizon={128 * 1024}}}"
+
+    env = neon_env_builder.init_start()
+
+    tenant_id = env.initial_tenant
+    main_branch_name, main_timeline_id = env.neon_cli.list_timelines(tenant_id)[0]
+
+    http_client = env.pageserver.http_client()
+
+    main_pg = env.postgres.create_start(main_branch_name, tenant_id=tenant_id)
+
+    batch_size = 10000
+
+    with main_pg.cursor() as cur:
+        cur.execute(
+            f"CREATE TABLE t0 AS SELECT i::bigint n FROM generate_series(0, {batch_size}) s(i)"
+        )
+
+    wait_for_last_flush_lsn(env, main_pg, tenant_id, main_timeline_id)
+    size_at_branch = get_tenant_size(http_client, tenant_id)[0]
+    assert size_at_branch > 0
+
+    first_branch_timeline_id = env.neon_cli.create_branch(
+        "first-branch", main_branch_name, tenant_id
+    )
+
+    # unsure why this happens
```

Comment:
That would be interesting to know indeed, we have crated two branches, so we can check the size differences and guess better. Maybe the difference is not even a constant?

Reply:
Got interested myself

```
diff --git a/test_runner/regress/test_tenant_size.py b/test_runner/regress/test_tenant_size.py
index 25b0f478..d0690c5c 100644
--- a/test_runner/regress/test_tenant_size.py
+++ b/test_runner/regress/test_tenant_size.py
@@ -194,6 +194,7 @@ def test_get_tenant_size_with_multiple_branches(neon_env_builder: NeonEnvBuilder
     # unsure why this happens
     size_after_branch = get_tenant_size(http_client, tenant_id)[0]
     assert size_after_branch > size_at_branch
+    log.error(f'size_after_branch: {size_after_branch}, size_at_branch: {size_at_branch}, first diff: {size_after_branch - size_at_branch}')
 
     first_branch_pg = env.postgres.create_start("first-branch", tenant_id=tenant_id)
 
@@ -220,6 +221,7 @@ def test_get_tenant_size_with_multiple_branches(neon_env_builder: NeonEnvBuilder
     )
     size_after_branch = get_tenant_size(http_client, tenant_id)[0]
     assert size_after_branch > size_after_continuing_on_main
+    log.error(f'size_after_branch: {size_after_branch}, size_after_continuing_on_main: {size_after_continuing_on_main}, second diff: {size_after_branch - size_after_continuing_on_main}')
 
     second_branch_pg = env.postgres.create_start("second-branch", tenant_id=tenant_id)
 ```

and that shows the result

```
2022-11-02 22:01:16.430 ERROR [test_tenant_size.py:197] size_after_branch: 24556288, size_at_branch: 24425216, first diff: 131072
2022-11-02 22:01:19.218 ERROR [test_tenant_size.py:224] size_after_branch: 26719448, size_after_continuing_on_main: 26580104, second diff: 139344
```

the diff seems close, but odd that it's a bit different.
Also, I seem to have used proper variables to find the diff, is that `131072` _bytes_? Seems to be quite a difference..

Reply:
131072 with one of these tests sounds familiar, might be the gc horizon, or maybe there's some alignment going on with the wal? I didn't yet go through yesterdays successful testing run, which should have the outputs for single branch testing to compare.

Reply:
The value matches the gc_horizon. I'll push a commit with assertion, lets see if it's stable.

Reply:
It did pass on both debug and release. Something to think about in future.

Reply:
So, every branch that we create, increments the total `tenant_size` by `gc_horizon` value?
I would not expect this behavior, so seems like a bug? If you prefer to consider it later, let's add an issue or a TODO around.

----------
Maybe, due to `pricing` concept introduced initially, I'm more critical to such errors than usual: since it's the end user that get billed for these errors.
If I'm wrong with my assumptions, let's move on then.

Reply:
Noted in the TODO issue.